### PR TITLE
4-level JS namespace

### LINKED ISSUE #745: decide on 4-level JS namespace
see #622

A 4-level JS namespace would allow to clearly separate installable application scripts and the framework.

### LINKED ISSUE #956: org.phoenicis.scripts.interpreter.ScriptException: Script not found: [Functions, Engines, Wine]
https://pastebin.com/yEw6cJi8

This error comes up when trying to install any version of Wine in the "Engines" tab. Four error dialogs show up and this text is outputted to the terminal log.

Also, once this happens, the window becomes un-resizable, it is stuck at a fixed size.

My system: Xubuntu 16.04 LTS 64-bit, openjdk version "1.8.0_131" from the Ubuntu repos, PoL built using `mvn clean package` after downloading the repository and installing dependencies.

### PR DESCRIPTION
**Pull request by plata**
fixes #745 
fixes #956

### CODE REVIEW

**Comment by madoar** [comment]
It seems like the tests are failing?

File: phoenicis-repository/src/main/java/org/phoenicis/repository/dto/TypeDTO.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * Represents a repository type
```
Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/dto/TypeDTO.java]: Can you please also add some examples for repository types?
E.g. engines, applications...

File: phoenicis-repository/src/main/java/org/phoenicis/repository/dto/TypeDTO.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    private final List<CategoryDTO> categories;
+    private final URI icon;
+
+    private TypeDTO(Builder builder) {
+        if (builder.id != null) {
+            if (builder.id.matches("^[a-zA-Z0-9]+$")) {
+                this.id = builder.id;
+            } else {
+                LOGGER.warn(
+                        String.format("Category ID (%s) contains invalid characters, will remove them.", builder.id));
```
Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/dto/TypeDTO.java]: I believe you mean `Type ID` right?

**Comment by plata** [comment]
Didn't run `clean` locally...

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java [hunk 3]
```diff
@@ -39,7 +39,7 @@
 
 public class LocalRepository implements Repository {
     private final static Logger LOGGER = LoggerFactory.getLogger(LocalRepository.class);
-    private static final String CATEGORY_ICON_NAME = "icon.png";
+    private static final String ICON_NAME = "icon.png";
```
Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java]: With the reduced name this constant needs a description. E.g. what does the icon represent, currently I would assume it represents the repository?

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java [hunk 4]
```diff
@@ -239,6 +279,15 @@ public RepositoryDTO fetchInstallableApplications() {
         return results;
     }
 
+    private TypeDTO unSerializeType(File jsonFile) {
```
Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java]: I would prefer `unserializeType`, this looks more natural than `unSerializeType`

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/Repository.java [hunk 5]
```diff
@@ -56,12 +53,40 @@ default void fetchInstallableApplications(Consumer<RepositoryDTO> callback, Cons
         }
     }
 
+    default TypeDTO getType(List<String> path) {
```
Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/Repository.java]: JavaDoc please

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/Repository.java [hunk 6]
```diff
@@ -56,12 +53,40 @@ default void fetchInstallableApplications(Consumer<RepositoryDTO> callback, Cons
...
+        final Optional<TypeDTO> typeDTO = fetchInstallableApplications().getTypes().stream()
+                .filter(type -> path.get(0).equals(type.getId())).findFirst();
+        return typeDTO.orElse(null);
+    }
+
+    default CategoryDTO getCategory(List<String> path) {
```
Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/Repository.java]: JavaDoc please

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/Repository.java [hunk 7]
```diff
@@ -56,12 +53,40 @@ default void fetchInstallableApplications(Consumer<RepositoryDTO> callback, Cons
...
+        if (typeDTO == null) {
+            return null;
+        }
+
+        final Optional<CategoryDTO> categoryDTO = typeDTO.getCategories().stream()
+                .filter(category -> path.get(1).equals(category.getId())).findFirst();
+        return categoryDTO.orElse(null);
+    }
+
+    default ApplicationDTO getApplication(List<String> path) {
```
Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/Repository.java]: JavaDoc please

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java (hunk 4 above)
Comment (plata) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java]: Yes. But we should change that separately (I used the same name as in `CategoryDTO`).

Comment (madoar) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java]: That is ok for me

**Comment by plata** [comment]
@qparis could you please check as well?

**Comment by plata** [comment]
@qparis do you want to give some feedback? Otherwise I would just merge this for now. I think we already agreed on the general way this should look like in #954 and PlayOnLinux/Scripts#296 and we can still improve this later on.

**Pull request merged**